convert DecimalBlock value back to Decimal for #4647

yvess opened the pull request:
_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [X] Does the code comply with the style guide?
    -   [X] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

see issue #4647
DecimalBlock values work as expected when previewing a page, however they become strings when the published page is rendered.

A decimal value is stored as string in a JSON field. To get a real Decimal value back in python, it needs to be casted as Decimal. Without this PR you get a string type object back instead.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

lb- commented:
@yvess please can you advise if this is still a draft (if so, can you change it to a draft) otherwise can you remove the 'draft' from the title.

Also - can you add a reference to the bug that this fixes in the PR's description.

yvess commented:
> @yvess please can you advise if this is still a draft (if so, can you change it to a draft) otherwise can you remove the 'draft' from the title.
> 
> Also - can you add a reference to the bug that this fixes in the PR's description.

Updated the description and added test. Not sure if this should be document is some way.

gasman commented:
Thanks @yvess - good job on diagnosing and fixing this.

I'd just suggest this as an additional test, to confirm that this really does achieve the effect asked for in the original issue (namely, ensuring that the type is preserved on a round trip to the database format and back):

```python
    def test_round_trip_to_db_preserves_type(self):
        block = blocks.DecimalBlock()
        original_value = Decimal(1.63)
        db_value = json.dumps(block.get_prep_value(original_value), cls=DjangoJSONEncoder)
        restored_value = block.to_python(json.loads(db_value))
        self.assertEqual(type(restored_value), Decimal)
        self.assertEqual(original_value, restored_value)
```
I'll add this when I merge.

gasman commented:
Merged in c5c93cba67661fa6fbb922db6261c51a2e5b7c1c
